onnx/defs/math/defs.cc

```diff
@@ -1158,7 +1158,78 @@ ONNX_OPERATOR_SET_SCHEMA(
         .TypeConstraint(
             "T",
             OpSchema::all_tensor_types(),
-            "Constrain input and output types to all tensors."));
+            "Constrain input and output types to all tensors.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Type inference
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+          // Shape inference
+          // For shape inference (and rank inference), we need both input shape
+          // and values in 'shape' tensor
+          const auto* shape_initializer = ctx.getInputData(1);
+          if (hasNInputShapes(ctx, 1) && nullptr != shape_initializer) {
+            const auto& shape_initializer_shape =
+                ctx.getInputType(1)->tensor_type().shape();
+            if (shape_initializer_shape.dim_size() != 1 ||
+                shape_initializer->data_type() != TensorProto::INT64)
+              fail_shape_inference(
+                  "'shape' input must be 1D tensor of type INT64");
+
+            const auto& input_shape =
+                ctx.getInputType(0)->tensor_type().shape();
+            const auto& shape_data = ParseData<int64_t>(shape_initializer);
+
+            size_t max_dimension_count = std::max(
```

Comment:
is this following the same broadcasting as https://github.com/onnx/onnx/blob/master/onnx/defs/shape_inference.h#L435 please?

Reply:
@linkerzhang - indeed it is. Thanks for pointing me to these helper functions. Resolved.